Delete/unexport TransactionRestartError and IndexedError

kkaneda opened the pull request:
This is the first step for #2809 and #3303. There are several cleanup I still have to make, but I wanted to make sure the overall direction is good.

This commit contains the following changes:
- Add txn and index to Error. Use them whenever possible (e.g., `txnSender.Send`).
- Remove `index` from `WriteIntentError` and `ConditionFailedError`.
- Change `Replica.executeCommand` to return `roachpb.Error`. Change transitively affected code to use `roachpb.Error` and avoid calling `GoError()` as much as possible.

Note that `roachpb.Error` has _NOT_ yet implemented the `Error` interface. This is necessary (at least during transition) to avoid mixing `roachpb.Error` and other Go errors.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3661)

<!-- Reviewable:end -->

kkaneda committed:
Delete/unexport TransactionRestartError and IndexedError

This is the first step for #2809 and #3303. There are several cleanup
I still have to make, but I wanted to make sure the overall direction is
good.

This commit contains the following changes:

- Add txn and index to Error. Use them whenever possible (e.g., txnSender.Send).
- Remove index from WriteIntentError and ConditionFailedError.
- Change Replica.executeCommand to return roachpb.Error. Change
  transitively affected code to use roachpb.Error and avoid
  calling GoError() as much as possible.

Note that roachpb.Error has *NOT* yet implemented the Error interface.
This is necessary (at least during transition) to avoid mixing
roachpb.Error and other Go errors.

tschottdorf commented:
Reviewed 100 of 100 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 123 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RgTls_XR492ofZ-jv:-K7Rg3TUyKUZQSmDzhBb:-172972443) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/client/txn.go#L123)):*
why is this a user-facing error? It would possibly happen to kv users, but those are all internal. SQL doesn't even give you the vocabulary to change the level during a transaction, so I don't think this is a "user error". Comment possibly applies in other locations as well, but I may have missed them.

---

*[roachpb/errors.go, line 73 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7Ri9kv8ikEQ9ItcK2E:-K7Ri9kw7PwpN4wUkm0c:437185678) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/roachpb/errors.go#L73)):*
why `2`?

---

*[roachpb/errors.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RiFxTUiJ8dJ0RRhTH:-K7RiFxTUiJ8dJ0RRhTI:404057559) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/roachpb/errors.go#L84)):*
the original reason for this type was to be able to wrap errors which weren't represented in `ErrorDetail` as a `Error` and have `GoError()` return something useful. Why does this need to be exported?

---

*[roachpb/errors.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7Rijk-rfwQAXCBTS7k:-K7Rijk-rfwQAXCBTS7l:-847647519) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/roachpb/errors.go#L141)):*
I think you could've just unexported the `TransactionRestartError` interface (and its contained method) and kept using it. After all, all it would do is assign restart behaviour to errors.

---

*[storage/replica.go, line 1465 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RlNnmukjFFCQJnyny:-K7RlNnmukjFFCQJnynz:-895686192) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica.go#L1465)):*
add a TODO here that this is prone to code rot. We should always set the index when the error stems from an individual command.

---

*[storage/replica.go, line 1740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RlwdaK3TmWKGHjxHC:-K7RlwdaK3TmWKGHjxHD:-792568857) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica.go#L1740)):*
just remove `chainedError` and manually create an error string in this method.

---

*[storage/replica.go, line 1865 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7Rm5pSMPuMkpoMB1Va:-K7Rm5pSMPuMkpoMB1Vb:-767949930) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica.go#L1865)):*
use `roachpb.NewErrorf("....")`

---

*[storage/replica_command.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RmEGQx6z4TrCOfRZ4:-K7RmEGQx6z4TrCOfRZ5:-1972282832) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica_command.go#L53)):*
was this important somewhere or just for completeness?

---

*[storage/replica_command.go, line 199 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RmNgBBMQ7Hs5-sFxF:-K7RmNgBBMQ7Hs5-sFxG:-1089886480) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica_command.go#L199)):*
why don't you just put `respHeader := reply.Header()` here instead of copying it in each single case above?

---

*[storage/replica_command.go, line 899 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7Rmu_yao9rRaRxlt_6:-K7Rmu_yao9rRaRxlt_7:640531389) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica_command.go#L899)):*
`if len(args.PusherTxn.ID) == 0`

---

*[storage/store.go, line 1469 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RncmQr5sAQSc05Sgw:-K7RncmQr5sAQSc05Sgx:1506323485) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/store.go#L1469)):*
what's this about? Maybe you're wondering why we're not setting the transaction in the header? The Push doesn't happen as official part of the transaction and this is by choice (there's an issue for it with discussion somewhere).

---

*[storage/stores.go, line 137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RoEjwmyMMy08M7dhO:-K7RoEjwmyMMy08M7dhP:-2012084445) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/stores.go#L137)):*
this should be gone before merge.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Geez, that must've been a horrible change to make. Some remarks on top of my inline comments.

I like the general direction and I would hold off on having `*Error` implement `error` again (we might want to someday, but it doesn't have any obvious benefits other than possibly being more idiomatic and with slightly more tooling support, but comes with a number of potential traps).

There are numerous variables called `err` but of type `roachpb.Error` in the code; i'd prefer to avoid `err` unless we implement `error`. `pErr` is used mostly (I realize that you were likely fixing that anyway).

What are your plans for going forward? I suppose the eventual goal is to un-implement `error` from every `ErrorDetail` member, and to make `GoError()` dumb (basically `errors.New(pErr.Error())`), which amounts to making sure the `util.Retryable` interface is never applied to it.

tschottdorf commented:
Oh, and thanks for tackling this nasty beast, @kkaneda!

kkaneda commented:
Thank you for the quick review, @tschottdorf ! 

> There are numerous variables called err but of type roachpb.Error in the code; i'd prefer to avoid err unless we implement error. pErr is used mostly (I realize that you were likely fixing that anyway).

Yeah, makes sense. Will change to pErr whenever possible. I kept `err`, but it's been confusing. :)

> What are your plans for going forward? I suppose the eventual goal is to un-implement error from every ErrorDetail member, and to make GoError() dumb (basically errors.New(pErr.Error())), which amounts to making sure the util.Retryable interface is never applied to it.

Honestly, I haven't thought much, but yes, un-implementing error from every `ErrorDetail` sounds like the right direction. Will take a look more after this PR.

Will address the comments and also make some more cleanup (and fix broken tests).

tamird commented:
Before more work goes into this, here's one important thing to remember: moving away from implementing `error` means `errcheck` can't help us make sure we're properly handling errors.

Perhaps it's time to write a new linter called `returncheck`.

If only we had something like [`#[must_use]`](https://github.com/rust-lang/rust/blob/master/src/test/rustdoc/must-use.rs#L12).

kkaneda commented:
Ah, very good point. 

Looking at https://github.com/kisielk/errcheck/blob/master/internal/errcheck/errcheck.go, and it doesn't like too much work needed to implement a new linter. I can work on after finishing this PR.

kkaneda commented:
Review status: 19 of 105 files reviewed at latest revision, 12 unresolved discussions.

---

*[client/txn.go, line 123 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RgTls_XR492ofZ-jv:-K7Ulrv3QpqQxtlPgq9h:2068697648) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/client/txn.go#L123)):*
I see. I simply used `NewUErrorf` whenever the original code uses `fmt.Errorf`. Fixed.

---

*[roachpb/errors.go, line 73 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7Ri9kv8ikEQ9ItcK2E:-K7Uhc7M1UCEIq_3T_b2:-1720606628) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/roachpb/errors.go#L73)):*
This was badly copied from `util.defaultSkip` in `util/error.go`... Fixed to 1 as we don't need to skip two stackframes (`getPrefix` and `Errorf`).

---

*[roachpb/errors.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RiFxTUiJ8dJ0RRhTH:-K7UhQXdFaOhTp-_8xQp:1566187469) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/roachpb/errors.go#L84)):*
This was only needed by `TxnCoordSender.updateStat` to run the following check:

```
	case *roachpb.InternalError:
		if pErr.Detail != nil {
			panic(fmt.Sprintf("unhandled InternalError %T", pErr))
		}
```

The check is somewhat redundant, I think. Removed and unexposed `InternalError.

---

*[roachpb/errors.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7Rijk-rfwQAXCBTS7k:-K7Um8giCphLpfzgNhjt:1931465779) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/roachpb/errors.go#L141)):*
Ah, good idea! Fixed.

---

*[storage/replica.go, line 1465 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RlNnmukjFFCQJnyny:-K7Up1EFzxUsuKfsB8S1:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica.go#L1465)):*
Done.

---

*[storage/replica.go, line 1740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RlwdaK3TmWKGHjxHC:-K7Uvo5qvQiZYoAvjdbw:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica.go#L1740)):*
Done.

---

*[storage/replica.go, line 1865 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7Rm5pSMPuMkpoMB1Va:-K7UwANXspW9OYuL__yy:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica.go#L1865)):*
Done.

---

*[storage/replica_command.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RmEGQx6z4TrCOfRZ4:-K7WfVjeuvfYEwbaIiiY:333267311) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica_command.go#L53)):*
This was just for completeness. No test failed without this change.

---

*[storage/replica_command.go, line 199 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RmNgBBMQ7Hs5-sFxF:-K7WfYtlo-wKhl6jpn_O:-233250190) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica_command.go#L199)):*
Ah, you're totally right.. Fixed.

---

*[storage/replica_command.go, line 899 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7Rmu_yao9rRaRxlt_6:-K7UwUFSs1YEAQkHAu-A:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/replica_command.go#L899)):*
Done.

---

*[storage/store.go, line 1469 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RncmQr5sAQSc05Sgw:-K7Uxusgnx8S0hpT-XsX:-968884310) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/store.go#L1469)):*
Ah, yes. I was wondering that as we might not set a txn for an `roachpb.Error` returned by `resolveWriteIntentError`.

---

*[storage/stores.go, line 137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RoEjwmyMMy08M7dhO:-K7UwVu7SGF6EzIfkYcf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/stores.go#L137)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Updated the PR. @tschottdorf , could you take another look? Thanks!

tschottdorf commented:
Reviewed 31 of 31 files at r2, 3 of 3 files at r3, 5 of 5 files at r4, 73 of 73 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[kv/split_test.go, line 255 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a6ZyJd50jq922Ijkz:-K7a6ZyKZk1Oirk5kzXu:1849609490) ([raw file](https://github.com/cockroachdb/cockroach/blob/9cef6ece03a0c8c810ee2ae707589204a26c20ef/kv/split_test.go#L255)):*
while you're here remove that. We've decided to just time out in cases like that.

---

*[roachpb/errors.go, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a2xFHDlUzTD68bL28:-K7a2xFI_tS0i5ZbA2RY:1930630722) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/roachpb/errors.go#L44)):*
I pictured this unexported as well for this to make sense. Is that not possible?

---

*[roachpb/errors.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a4-1g9Fq5jj-5yRjN:-K7a4-1g9Fq5jj-5yRjO:-203227068) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/roachpb/errors.go#L325)):*
the interface name is now lowercase. Ditto above.

---

*[sql/scan.go, line 530 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a45two3XBtOodcUxD:-K7a45two3XBtOodcUxE:1313663657) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/sql/scan.go#L530)):*
what's this about?

---

*[sql/sort.go, line 91 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a48Xtt9M1U7oPp8_i:-K7a48Xtt9M1U7oPp8_j:-2088511885) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/sql/sort.go#L91)):*
and this?

---

*[storage/replica.go, line 1761 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a54jfp2yBwAbdHt6s:-K7a54jgRzx50CEazCJZ:-1209564082) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/storage/replica.go#L1761)):*
lose the `!= nil` check (convention).

---

*[storage/store.go, line 1469 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RncmQr5sAQSc05Sgw:-K7a5O_zF835Bg3zvNT3:538678614) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/store.go#L1469)):*
hmm, that's a good point. Add a TODO.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
LGTM. With the amount of code changes, you're going to be fixing your commits after each rebase as others merge their work, so let's do a quick a) squash b) rebase c) run CI 5x d) merge.
PS Just to make sure, you were using `gorename`, right? :-)

kkaneda commented:
Sure! Will do final rebase and run CI tonight.

Seriously, I didn't know `gorename`. :) Thanks for the info!

Thanks again!

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[kv/split_test.go, line 255 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a6ZyJd50jq922Ijkz:-K7l28rBNvzpjrwmflrN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9cef6ece03a0c8c810ee2ae707589204a26c20ef/kv/split_test.go#L255)):*
Done.

---

*[roachpb/errors.go, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a2xFHDlUzTD68bL28:-K7l2KkRpGzlDlMXheNS:-316759653) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/roachpb/errors.go#L44)):*
Ah, no.. It's doable. Fixed.

---

*[roachpb/errors.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a4-1g9Fq5jj-5yRjN:-K7l2V65kN-iy5Y58YOH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/roachpb/errors.go#L325)):*
Done.

---

*[sql/scan.go, line 530 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a45two3XBtOodcUxD:-K7l8-43OeUHM3JmIS_b:-331552661) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/sql/scan.go#L530)):*
Ah, right. We don't need to change this. Fixed.

---

*[sql/sort.go, line 91 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a48Xtt9M1U7oPp8_i:-K7l4qzWYGnBLDCJkQd6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/sql/sort.go#L91)):*
Done.

---

*[storage/replica.go, line 1761 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a54jfp2yBwAbdHt6s:-K7l8252LGgCRMyOpPP_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/storage/replica.go#L1761)):*
Done.

---

*[storage/store.go, line 1469 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7RncmQr5sAQSc05Sgw:-K7l5S_8Jh7EmvGyQwoD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9364ce853059d568ccfa9114c8ec742718263c98/storage/store.go#L1469)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-:-K7l85jKFCa0FNSZXI8m:-1745131173)*
<!-- Sent from Reviewable.io -->

kkaneda commented on sql/lease.go:
```diff
@@ -272,8 +278,8 @@ func (s LeaseStore) Publish(tableID ID, update func(*TableDescriptor) error) err
 
-		if err != errLeaseVersionChanged {
-			return err
+		if _, ok := pErr.GoError().(*errLeaseVersionChanged); !ok {
```
This is not working.. Will fix.

kkaneda replied:
I've ended up adding more messages to error.proto to be able to convert between go Errors and  proto messages. Added a TODO to revisit the design.

tschottdorf commented:
Reviewed 14 of 14 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/replica.go, line 1761 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a54jfp2yBwAbdHt6s:-K7lVzcdiX6b5heKKQrb:1100753763) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/storage/replica.go#L1761)):*
sorry, I meant you don't have to check for `!= nil` at all. If `pErr` is `nil`, then `.GoError()` is nil and `nil` doesn't type assert to `*roachpb.ReplicaCorruptionError`.
You only lose the case in which `GoError()` returns an `interface{}(*roachpb.ReplicaCorruptionError(nil))` and we've decided that that counts as a real error (by convention, not that we ever want such a thing to exist), so that's ok.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Review status: 70 of 105 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/replica.go, line 1761 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7a54jfp2yBwAbdHt6s:-K7njXcnre1qsuz17tBW:-119087499) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc4da2b63c907b1067e633f8163e0bdcb5922de/storage/replica.go#L1761)):*
Ah, thanks for the explanation! Fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Looking into test failures. The failure on TestStoreRangeSystemSplits seems like a known issue, but the one on TestAsyncSchemaChanger might be new.

tschottdorf commented:
I think that test is flaky in general. If you see only those two failures, safe to assume it's not introduced here.

tschottdorf commented:
LGTM. Merge when you get a chance. The rebasing will never stop.

---

Reviewed 46 of 46 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/backfill.go, line 70 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7p7J6Qvy1L87HYY-y4:-K7p7J6Qvy1L87HYY-y5:-1306278538) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f54ea313d1beb1c546e6b806d2b8bde7dd1f19a/sql/backfill.go#L70)):*
wonder why this is in your diff. I'm going to assume Reviewable got confused with the base commit changes.

---

*[sql/schema_changer.go, line 34 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7p7kUxpWnF0iRww1sw:-K7p7kUxpWnF0iRww1sx:1784984129) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f54ea313d1beb1c546e6b806d2b8bde7dd1f19a/sql/schema_changer.go#L34)):*
uncomment

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-:-K7p8UzvYNPkW-xpGibF:765310999)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Sure. Will try this morning or night.

---

Review status: 97 of 105 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/backfill.go, line 70 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7p7J6Qvy1L87HYY-y4:-K7q1YITD8EYdexOL-JC:-1013918202) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f54ea313d1beb1c546e6b806d2b8bde7dd1f19a/sql/backfill.go#L70)):*
Hmm, wasn't totally sure which change you're referring, but could be. The change I  made here is to just return `roachpb.Error` as the method calls `p.txn.Run`, which returns `roachpb.Error`. 

---

*[sql/schema_changer.go, line 34 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-K7p7kUxpWnF0iRww1sw:-K7q1XQoEmJHlS7ARjXF:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f54ea313d1beb1c546e6b806d2b8bde7dd1f19a/sql/schema_changer.go#L34)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3661#-:-K7qCwi368-X1N-ZAWB0:580987676)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Merging.
